projects/online/online/main.py

```diff
@@ -245,26 +192,136 @@ def search(
+    rejected_path: Path,
+    astro_event_rate: float,
+    outdir: Path,
+):
+    gdb = gracedb_factory(server, outdir)
+
+    logging.info("Fitting p_astro model or loading from cache")
+    pastro_model = fit_or_load_pastro(
+        outdir / "pastro.pkl",
+        background_path,
+        foreground_path,
+        rejected_path,
+        astro_event_rate=astro_event_rate,
+    )
+    logging.info("Loaded p_astro model")
+
+    while True:
+        event = pastro_queue.get()
+        logging.info("Calculating p_astro")
+        pastro = pastro_model(event.detection_statistic)
+        graceid = pastro_queue.get()
+        logging.info(f"Submitting p_astro: {pastro}")
+        gdb.submit_pastro(float(pastro), graceid, event.gpstime)
+        logging.info("Submitted p_astro")
+
+
+def amplfi_subprocess(
+    amplfi_queue: Queue,
+    server: GdbServer,
+    outdir: Path,
+    inference_params: List[str],
+    shared_samples: Array,
+    nside: int = 32,
+):
+    gdb = gracedb_factory(server, outdir)
+
+    while True:
+        arg = amplfi_queue.get()
+        if isinstance(arg, float):
+            event_time = arg
+            descaled_samples = torch.reshape(
+                torch.Tensor(shared_samples), (-1, len(inference_params))
+            )
+            logging.info("Creating skymap")
+            posterior, mollview_map, skymap = skymap_from_samples(
+                descaled_samples, event_time, inference_params, nside
+            )
+            graceid = amplfi_queue.get()
+            logging.info("Submitting PE")
+            gdb.submit_pe(posterior, mollview_map, skymap, graceid, event_time)
+            logging.info("Submitted all PE")
+        else:
+            graceid = arg
+            event_time = amplfi_queue.get()
+            descaled_samples = torch.reshape(
+                torch.Tensor(shared_samples), (-1, len(inference_params))
+            )
+            logging.info("Creating skymap")
+            posterior, mollview_map, skymap = skymap_from_samples(
+                descaled_samples, event_time, inference_params, nside
+            )
+            logging.info("Submitting PE")
+            gdb.submit_pe(posterior, mollview_map, skymap, graceid, event_time)
+            logging.info("Submitted all PE")
+
+
+def event_creation_subprocess(
+    event_queue: Queue,
+    server: GdbServer,
+    outdir: Path,
+    amplfi_queue: Queue,
+    pastro_queue: Queue,
+):
+    gdb = gracedb_factory(server, outdir)
+    last_auth = time.time()
+    while True:
+        try:
+            event = event_queue.get_nowait()
+            logging.info("Putting event in pastro queue")
+            pastro_queue.put(event)
+
+            # write event information to disk
+            # and submit it to gracedb
+            event.write(outdir)
+            response = gdb.submit(event)
+            # Get the event's graceid for submitting
+            # further data products
+            if isinstance(response, str):
```

Comment:
would it make more sense to do the check on the `server` variable instead of the type of the respoinse?

Reply:
Yeah, I think that would be clearer